lib/ecto/repo/schema.ex

```diff
@@ -42,21 +42,26 @@ defmodule Ecto.Repo.Schema do
       |> returning(opts)
       |> fields_to_sources(dumper)
 
-    {rows, header, placeholder_values} =
-      extract_header_and_fields(rows, schema, dumper, autogen_id, placeholder_map, adapter)
+    {rows_or_query, header, placeholder_values} =
+      extract_header_and_fields(rows_or_query, schema, dumper, autogen_id, placeholder_map, adapter)
+
+    counter =
+      case rows_or_query do
+        rows when is_list(rows) -> fn -> Enum.reduce(rows, 0, &length(&1) + &2) end
+        {%Ecto.Query{}, params} -> fn -> length(params) end
+      end
```

Comment:
we should likely move the counter to `extract_header_and_fields` and keep the logic all encapsulated in one place. :)

Reply:
ok!